Added AllocMonitor facility

Dr15Jones opened the pull request:
#### PR description:

This is a general purpose, extensible memory allocation/deallocation monitor systems. See README.md files for details.

#### PR validation:

Code compiles and run on x86 with new unit tests passing.

Dr15Jones committed:
Added AllocMonitor facility

A general system to watch allocations/deallocations

Dr15Jones commented:
assign core

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42665/36736

- This PR adds an extra 24KB to repository

cmsbuild commented:
New categories assigned: core

@Dr15Jones,@smuzaffar,@makortel you have been requested to review this Pull request/Issue and eventually sign? Thanks

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

- PerfTools/AllocMonitor (****)
- PerfTools/AllocMonitorPreload (****)
- PerfTools/MaxMemoryPreload (****)


The following packages do not have a category, yet:

PerfTools/AllocMonitor
PerfTools/AllocMonitorPreload
PerfTools/MaxMemoryPreload
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please review it and eventually sign? Thanks.
@makortel, @felicepantaleo this is something you requested to watch as well.
@perrotta, @dpiparo, @antoniovilela, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
@dan131riley would it be possible for you to try this on ARM? from compiler explorer the mangled names are different. It would be nice to see if we can make this work on ARM as well.

Dr15Jones commented:
please test

Dr15Jones commented:
I updated the proxy test to test all the `operator new` variations. AFAIK there is no way to force a version of `operator delete`, that seems to be up to the compiler to choose.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42665/36737

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #42665 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d3bb6b/34495/summary.html
**COMMIT**: 6d71186162288987f549b4fba6d1e16e44ee161a
**CMSSW**: CMSSW_13_3_X_2023-08-25-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/42665/34495/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 19 lines from the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3153095
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3153070
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-08-25-1100+d3bb6b/58613/triggerResults): no differences found

dan131riley commented:
> @dan131riley would it be possible for you to try this on ARM? from compiler explorer the mangled names are different. It would be nice to see if we can make this work on ARM as well.

Doesn't even make it to the name mangling part, looks to get tied up in recursive calls to cmalloc() while trying to hook malloc():

```
(gdb) bt
#0  0x0000ffff87279210 in syscall () from /lib64/libc.so.6
#1  0x0000ffff874a2204 in __cxxabiv1::__cxa_guard_acquire (g=0xffff8793a0b0 <guard variable for calloc::original>) at ../../../../libstdc++-v3/libsupc++/guard.cc:319
#2  0x0000ffff8791b574 in calloc (nitems=1, item_size=32) at /home/dsr/CMSSW_13_3_0_pre2/src/PerfTools/AllocMonitorPreload/src/memory_proxies.cc:45
#3  0x0000ffff87723890 in _dlerror_run () from /lib64/libdl.so.2
#4  0x0000ffff87723278 in dlsym () from /lib64/libdl.so.2
#5  0x0000ffff8791d928 in (anonymous namespace)::get<void* (*)(long unsigned int, long unsigned int) noexcept>(const char *) (iName=0xffff8791e4d0 "calloc") at /home/dsr/CMSSW_13_3_0_pre2/src/PerfTools/AllocMonitorPreload/src/memory_proxies.cc:19
#6  0x0000ffff8791b594 in calloc (nitems=1, item_size=32) at /home/dsr/CMSSW_13_3_0_pre2/src/PerfTools/AllocMonitorPreload/src/memory_proxies.cc:45
#7  0x0000ffff87723890 in _dlerror_run () from /lib64/libdl.so.2
#8  0x0000ffff87723278 in dlsym () from /lib64/libdl.so.2
#9  0x0000ffff8791d8d4 in (anonymous namespace)::get<void* (*)(long unsigned int) noexcept>(const char *) (iName=0xffff8791e4c8 "malloc") at /home/dsr/CMSSW_13_3_0_pre2/src/PerfTools/AllocMonitorPreload/src/memory_proxies.cc:19
#10 0x0000ffff8791b3d4 in malloc (size=72704) at /home/dsr/CMSSW_13_3_0_pre2/src/PerfTools/AllocMonitorPreload/src/memory_proxies.cc:35
#11 0x0000ffff8749dfb4 in (anonymous namespace)::pool::pool (this=0xffff876023b8 <(anonymous namespace)::emergency_pool>) at ../../../../libstdc++-v3/libsupc++/eh_alloc.cc:123
#12 __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at ../../../../libstdc++-v3/libsupc++/eh_alloc.cc:262
#13 _GLOBAL__sub_I_eh_alloc.cc(void) () at ../../../../libstdc++-v3/libsupc++/eh_alloc.cc:338
#14 0x0000ffff8794a22c in call_init.part () from /lib/ld-linux-aarch64.so.1
#15 0x0000ffff8794a330 in _dl_init () from /lib/ld-linux-aarch64.so.1
#16 0x0000ffff8793d46c in _dl_start_user () from /lib/ld-linux-aarch64.so.1
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb)
```

Dr15Jones commented:
@dan131riley Thanks for giving it a try. IF I found the right version of the code and IF I have interpreted it correctly, it looks like the dlsym is trying to do a one time setup of space for this thread's error reporting. It might be possible to identify that case and handle it by returning a previously _file scope static_ allocated memory block. Would mean `free` would have to watch out for that address as well.

dan131riley commented:
@Dr15Jones some prior art:

- https://stackoverflow.com/questions/17803456/an-alternative-for-the-deprecated-malloc-hook-functionality-of-glibc
- https://stackoverflow.com/questions/6083337/overriding-malloc-using-the-ld-preload-mechanism
- https://www.slideshare.net/tetsu.koba/tips-of-malloc-free (slide 35)

Generally seems to be having a special memory allocator just for that one allocation.  If we were using the glibc allocator all the time, we could call `__libc_calloc()` instead of getting it from `dlsym()`, but that won't work with (je|tc)malloc.

dan131riley commented:
@Dr15Jones I hacked together various bits and pieces at [dan131riley:AllocMonitor](https://github.com/dan131riley/cmssw/blob/AllocMonitor/PerfTools/AllocMonitorPreload/src/memory_proxies.cc#L10-L25), which seems to work, or at least passes tests.  I would have been surprised if the mangling were actually different--AFAIK, gcc and llvm follow the[ itanium ABI for mangling](https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling) on all platforms.

Dr15Jones commented:
@dan131riley Thanks for exploring further! The only issue I have with your implementation is it doesn't guarantee alignment rules if there is more than one allocation request to the local_* functions.

dan131riley commented:
@Dr15Jones I've updated to fix the alignment issue, good catch.  At least on ARM, there are two allocations--the first one is for an exception handling emergency pool, and the allocation for that then triggers the call to dlsym() that results in the second allocation.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42665/36757

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #42665 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42665/36759

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #42665 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

Commit:
make startup allocations safe on platforms where dlsym allocates memory

Dr15Jones committed:
Merge pull request #10 from dan131riley/AllocMonitorARM

make startup allocations safe on platforms where dlsym allocates memory

Dr15Jones commented:
please test

cmsbuild commented:
Pull request #42665 was updated. @makortel, @smuzaffar, @Dr15Jones can you please check and sign again.

Dr15Jones committed:
Add tests using jemalloc and tcmalloc

Dr15Jones commented on PerfTools/AllocMonitor/test/BuildFile.xml:
done

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42665/36772

- This PR adds an extra 16KB to repository

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d3bb6b/34523/summary.html
**COMMIT**: 962b9354d8bcec385859eb862a37d278d68adc04
**CMSSW**: CMSSW_13_3_X_2023-08-29-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/42665/34523/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 13 lines from the logs
 - Reco comparison results: 333 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3153095
 - DQMHistoTests: Total failures: 119
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3152954
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-08-29-1100+d3bb6b/58639/triggerResults): no differences found

makortel commented on PerfTools/AllocMonitorPreload/src/memory_proxies.cc:
```diff
@@ -0,0 +1,419 @@
+  // before we've finished our setup.  In particular, this avoids a
+  // chicken/egg problem if dlsym() allocates any memory.
+  constexpr auto max_align = alignof(std::max_align_t);
+  alignas(max_align) char tmpbuff[131072];
```
Could we add some comment on the size of the buffer (128 kB), like is it based on some estimation, or a guess that seems to work? (thinking a future reader that wonders why the size was set to 128 kB)

dan131riley replied:
The libstdc++ exception handling reserves an emergency buffer, which on ARM64 defaults to 72704 bytes.  That triggers a call to dlsym() to load malloc(), which allocates another 32 bytes for the error handler buffer.  So I took 72736 and rounded up to the next power of 2 to be on the safe side.

makortel commented on PerfTools/AllocMonitorPreload/src/memory_proxies.cc:
```diff
@@ -0,0 +1,419 @@
+}
+
+void* aligned_alloc(size_t alignment, size_t size) noexcept {
+  static auto original = get<decltype(&::aligned_alloc)>("aligned_alloc");
```
```suggestion
  static const auto original = get<decltype(&::aligned_alloc)>("aligned_alloc");
```
?

makortel commented on PerfTools/AllocMonitorPreload/src/memory_proxies.cc:
```diff
@@ -0,0 +1,419 @@
+}
+
+void free(void* ptr) noexcept {
+  static auto original = get<decltype(&::free)>("free");
```
```suggestion
  static const auto original = get<decltype(&::free)>("free");
```
?

makortel commented on PerfTools/AllocMonitorPreload/src/memory_proxies.cc:
```diff
@@ -0,0 +1,419 @@
+#include <new>
+
+void* operator new(std::size_t size) {
+  static auto original = get<void* (*)(std::size_t)>("_Znwm");
```
```suggestion
  static const auto original = get<void* (*)(std::size_t)>("_Znwm");
```
and hereafter (I won't repeat the comment more)?

makortel commented on PerfTools/AllocMonitorPreload/src/memory_proxies.cc:
```diff
@@ -0,0 +1,419 @@
+  }
+  auto& reg = AllocMonitorRegistry::instance();
+  reg.deallocCalled([ptr, al, &tag]() { original(ptr, al, tag); }, [ptr]() { return malloc_usable_size(ptr); });
+}  //_ZdlPvSt11align_val_tRKSt9nothrow_t
+void operator delete[](void* ptr, std::align_val_t al, const std::nothrow_t& tag) noexcept {
```
```suggestion
}  //_ZdlPvSt11align_val_tRKSt9nothrow_t

void operator delete[](void* ptr, std::align_val_t al, const std::nothrow_t& tag) noexcept {
```

makortel commented on PerfTools/MaxMemoryPreload/README.md:
```diff
@@ -0,0 +1,29 @@
+- The maximum amount of _used_ allocated memory that was in use at one time.
+- The amount of _used_ memory allocated during the job that has yet to be reclaimed by calling deallocation.
+- Number of calls made to allocation functions.
+- Number of calls made to deallocation functions.
+This service is multi-thread safe. Note that when run multi-threaded the maximum reported value will vary from job to job.
```
```suggestion
- Number of calls made to deallocation functions.

This service is multi-thread safe. Note that when run multi-threaded the maximum reported value will vary from job to job.
```

Dr15Jones committed:
Changes based on pull request comments

- improved documentation and formatting
- improved const use
- match CMS coding standards
- resolved static analyzer warnings

Dr15Jones commented on PerfTools/AllocMonitor/plugins/HistogrammingAllocMonitor.cc:
```diff
@@ -0,0 +1,87 @@
+      auto log = edm::LogSystem("HistogrammingAllocMonitor");
+      log << "Memory Histogram"
+          << "\n   size                 allocated           deallocated"
+          << "\n                  requested      used          used";
```
the two words belong with `allocated`.

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42665/36776

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #42665 was updated. @makortel, @smuzaffar, @Dr15Jones can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d3bb6b/34535/summary.html
**COMMIT**: b4220bec793cb340ba35e1cefd09638b3e8279bf
**CMSSW**: CMSSW_13_3_X_2023-08-29-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/42665/34535/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 9 lines from the logs
 - Reco comparison results: 32 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3153095
 - DQMHistoTests: Total failures: 1496
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3151577
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-08-29-1100+d3bb6b/58649/triggerResults): no differences found

makortel commented:
Comparisons show https://github.com/cms-sw/cmssw/issues/39803

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42665/36778

- This PR adds an extra 20KB to repository

makortel commented:
Comparisons show https://github.com/cms-sw/cmssw/issues/39803 again, although to lesser extent. This time both baseline and PR tests were run on `Intel(R) Xeon(R) CPU E5-2683 v4`.

makortel commented:
+core

rappoccio commented:
+1

smuzaffar commented:
@Dr15Jones , should this work for CentOS7 ? The unit tests in IBs are failing for slc7 https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc11/CMSSW_13_3_X_2023-09-04-1100/unitTestLogs/PerfTools/AllocMonitor#/

Dr15Jones commented:
@smuzaffar the expectation was it should work for CentOS7.

dan131riley commented:
The failures are all timeouts, so guess it’s deadlocking somewhere?

Dr15Jones commented:
#42724 fixes the problem. sl7 needs Dan's fix as well.
